File: day5/day5.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const MaxInt = int((^uint(0) >> 1))

type Range struct {
	dst int
	src int
	len int
}

func (r Range) Map(n int) (int, bool) {
	if n < r.src || n > (r.src+r.len-1) {
		return 0, false
	}
	return r.dst - r.src + n, true
}

type Map struct {
	name   string
	ranges []Range
}

func (m *Map) Append(r Range) {
	m.ranges = append(m.ranges, r)
}

func (m *Map) Map(n int) int {
	for i := range m.ranges {
		x, ok := m.ranges[i].Map(n)
		if ok {
			return x
		}
	}
	return n
}

type Almanac struct {
	seeds []int
	maps  []*Map
}

func (a *Almanac) ParseFile(name string) error {
	f, err := os.Open(name)
	if err != nil {
		return err
	}
	defer f.Close()
	scanner := bufio.NewScanner(f)
	scanner.Scan()
	seedsStr := scanner.Text()
	seeds := strToInts(seedsStr[7:])
	a.seeds = seeds
	scanner.Scan()
	var m *Map
	for scanner.Scan() {
		s := scanner.Text()
		s = strings.Trim(s, " \n")
		if s == "" {
			a.maps = append(a.maps, m)
			continue
		}
		if s[0] > '9' {
			m = &Map{
				name:   s,
				ranges: []Range{},
			}
			continue
		}
		nums := strToInts(s)
		r := Range{
			dst: nums[0],
			src: nums[1],
			len: nums[2],
		}
		m.Append(r)
	}
	a.maps = append(a.maps, m)
	return nil
}

func (a *Almanac) seedToLoc(n int) int {
	for i := range a.maps {
		n = a.maps[i].Map(n)
	}
	return n
}

func (a *Almanac) ClosestForSeeds() int {
	result := MaxInt
	for _, n := range a.seeds {
		loc := a.seedToLoc(n)
		if loc < result {
			result = loc
		}
	}
	return result
}

func (a *Almanac) ClosestForRanges() int {
	result := MaxInt
	for i := 0; i < (len(a.seeds) - 2); i += 2 {
		for j := a.seeds[i]; j < (a.seeds[i] + a.seeds[i+1]); j++ {
			loc := a.seedToLoc(j)
			if loc < result {
				result = loc
			}
		}
	}
	return result
}

func strToInts(s string) []int {
	s = strings.Trim(s, " \n")
	split := strings.Split(s, " ")
	nums := []int{}
	for _, x := range split {
		n, err := strconv.Atoi(x)
		if err != nil {
			panic(err)
		}
		nums = append(nums, n)
	}
	return nums
}

func main() {
	a := new(Almanac)
	a.ParseFile("input.txt")

	x := a.ClosestForSeeds()
	fmt.Println(x)

	y := a.ClosestForRanges()
	fmt.Println(y)
}

File: day5/day5.py
import sys
sys.path.append('../lib')

import tools

class Range(object):
    def __init__(self, dst_start, src_start, lenght):
        self.dst = dst_start
        self.src = src_start
        self.len = lenght

    def __str__(self):
        return f'range: {self.dst}, {self.src}, {self.len}'

    def map(self, n: int) -> int:
        if n < self.src or n > (self.src + self.len - 1):
            return None
        return self.dst - self.src + n   


class Map(object):
    def __init__(self, name):
        self.name = name
        self.ranges = []

    def __str__(self):
        out = f'map "{self.name}"\n'
        for r in self.ranges:
            out += str(r) + '\n'
        return out

    def append_range(self, r):
        self.ranges.append(r)

    def map(self, inp: int) -> int:
        for r in self.ranges:
            mapped = r.map(inp)
            if mapped is not None:
                return mapped
        return inp
            

class Almanac(object):
    def __init__(self, lines):
        self.seeds = [int(n) for n in lines[0].replace('seeds: ', '').split()]
        self.maps = list()
        m = None
        for line in lines[2:]:
            if line == '':
                self.maps.append(m)
                continue
            if line[0].isalpha():
                m = Map(line)
                continue
            a, b, c = [int(n) for n in line.split()]
            r = Range(a, b, c)
            m.append_range(r)
        self.maps.append(m)

    def __str__(self):
        out = 'The latest Island Island Almanac\n'
        out += f'Seeds: {", ".join([str(n) for n in self.seeds])}\n'
        for m in self.maps:
            out += str(m)
        return out

    def _map_seed_to_location(self, seed: int) -> int:
        for m in self.maps:
            seed = m.map(seed)
        return seed

    def find_closest_location_for_seed_nums(self) -> int:
        result = float('inf')
        for seed in self.seeds:
            loc = self._map_seed_to_location(seed)
            if loc < result:
                result = loc
        return result

    def find_closest_location_for_seed_ranges(self) -> int:
        closest = float('inf')
        for i in range(0, len(self.seeds), 2):
           for n in range(self.seeds[i], self.seeds[i]+self.seeds[i+1]):
                loc = self._map_seed_to_location(n)
                if loc < closest:
                    closest = loc
        return closest
        

if __name__ == '__main__':
    r = tools.Reader('input.txt')
    lines = r.read()

    al = Almanac(lines)

    x = al.find_closest_location_for_seed_nums()
    print(x)

    y = al.find_closest_location_for_seed_ranges()
    print(y)


    
